org.eclipse.jdt.ls.core/src/org/eclipse/jdt/ls/core/internal/handlers/BundleUtils.java

```diff
@@ -126,23 +140,55 @@ public static void loadBundles(Collection<String> bundleLocations) throws CoreEx
 				String location = getBundleLocation(bundleLocation, true);
 
 				BundleInfo bundleInfo = getBundleInfo(bundleLocation);
-				if (!bundleInfos.add(bundleInfo)){  // returns false if bundleInfo is not added because it already exists
-					continue;
-				 }
+
 				if (bundleInfo == null) {
 					status.add(new Status(IStatus.ERROR, context.getBundle().getSymbolicName(), "Failed to get bundleInfo for bundle from " + bundleLocation, null));
 					continue;
 				}
+
+				// Platform.getBundle() won't return the bundle with INSTALLED state, so we use this Map to persist the current installed bundles
+				List<Bundle> installedBundleList = installedBundles.get(bundleInfo.getSymbolicName());
+				if (installedBundleList != null) {
+					boolean shouldSkip = false;
+					for (Bundle bundle : installedBundleList) {
+						Module module = bundle.adapt(Module.class);
+						if (isSingleton(module.getCurrentRevision())) {

```

Comment:
The singleton information can be resolved from the method `getBundleInfo(String bundleLocation)` directly, no need to get it from the module revision.

Reply:
Do you mean getting it by parsing the symbol name? What if it contains whitespace?

Reply:
this is a sample format `Bundle-SymbolicName: org.eclipse.jdt.ls.core;singleton:=true` in MANIFEST.MF. Currently before installing the new bundle, we already has logic to resolve the symbolic name and version info from the manifest file directly, why not bring the singleton attribute by the way.

Reply:
Found a way to do it:

```
ManifestElement[] symbolicNameElements = ManifestElement.parseHeader(Constants.BUNDLE_SYMBOLICNAME, symbolicName);
if (symbolicNameElements.length > 0) {
    symbolicName = symbolicNameElements[0].getValue();
    String singleton = symbolicNameElements[0].getDirective(Constants.SINGLETON_DIRECTIVE);
    isSingleton = "true".equals(singleton);
}
```

Reply:
this is the existing utility method in BundleUtils. Looks like your sample code is better.
```java
	private static BundleInfo getBundleInfo(String bundleLocation) throws IOException {
		try (JarFile jarFile = new JarFile(bundleLocation)) {
			Manifest manifest = jarFile.getManifest();
			if (manifest != null) {
				Attributes mainAttributes = manifest.getMainAttributes();
				if (mainAttributes != null) {
					String bundleVersion = mainAttributes.getValue(Constants.BUNDLE_VERSION);
					if (StringUtils.isBlank(bundleVersion)) {
						return null;
					}
					String symbolicName = mainAttributes.getValue(Constants.BUNDLE_SYMBOLICNAME);
					if (StringUtils.isNotBlank(symbolicName) && symbolicName.indexOf(';') >= 0) {
						symbolicName = symbolicName.substring(0, symbolicName.indexOf(';'));
					}
					return new BundleInfo(bundleVersion, symbolicName);
				}
			}
		}
		return null;
	}
```